src/devices/src/virtio/block/device.rs

```diff
@@ -0,0 +1,285 @@
+            1 << VIRTIO_F_VERSION_1 | 1 << VIRTIO_F_IN_ORDER | 1 << VIRTIO_F_RING_EVENT_IDX;
+
+        if args.read_only {
+            device_features |= 1 << VIRTIO_BLK_F_RO;
+        }
+
+        // A block device has a single queue.
+        let queues = vec![Queue::new(args.common.mem, QUEUE_MAX_SIZE)];
+        let config_space = build_config_space(&args.file_path)?;
+        let virtio_cfg = VirtioConfig::new(device_features, queues, config_space);
+
+        // Used to send notifications to the driver.
+        let irqfd = EventFd::new(EFD_NONBLOCK).map_err(Error::EventFd)?;
+
+        args.common
+            .vm_fd
+            .register_irqfd(&irqfd, args.common.mmio_cfg.gsi)
+            .map_err(Error::RegisterIrqfd)?;
+
+        let mmio_cfg = args.common.mmio_cfg;
+
+        let block = Arc::new(Mutex::new(Block {
+            virtio_cfg,
+            mmio_cfg,
+            endpoint: args.common.event_mgr.remote_endpoint(),
+            vm_fd: args.common.vm_fd,
+            irqfd: Arc::new(irqfd),
+            file_path: args.file_path,
+            read_only: args.read_only,
+            _root_device: args.root_device,
+        }));
+
+        // Register the device on the MMIO bus.
+        args.common
+            .mmio_mgr
+            .register_mmio(mmio_cfg.range, block.clone())
+            .map_err(Error::Bus)?;
+
+        // TODO: We're hardcoding the MMIO range len to 4K so assert for this until we decide
+        // to do something else.
+        assert_eq!(mmio_cfg.range.size(), 0x1000);
+
+        // Extra parameters have to be appended to the cmdline passed to the kernel because
+        // there's no active enumeration/discovery mechanism for virtio over MMIO. In the future,
+        // we might rely on a device tree representation instead.
+
+        let root_str = if args.root_device {
+            " root=/dev/vda"
+        } else {
+            ""
+        };
+
+        args.common.kernel_cmdline.push_str(&format!(
+            " virtio_mmio.device=4K@0x{:x}:{}{}",
+            mmio_cfg.range.base().0,
+            mmio_cfg.gsi,
+            root_str,
+        ));
+
+        Ok(block)
+    }
+}
+
+// We now implement `WithVirtioConfig` and `WithDeviceOps` to get the automatic implementation
+// for `VirtioDevice`.
+impl<M: GuestAddressSpace + Clone + Send + 'static> WithVirtioConfig<M> for Block<M> {
+    fn device_type(&self) -> u32 {
+        BLOCK_DEVICE_ID
+    }
+
+    fn virtio_config(&self) -> &VirtioConfig<M> {
+        &self.virtio_cfg
+    }
+
+    fn virtio_config_mut(&mut self) -> &mut VirtioConfig<M> {
+        &mut self.virtio_cfg
+    }
+}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> WithDeviceOps for Block<M> {
+    type E = Error;
+
+    fn activate(&mut self) -> Result<()> {
+        if self.virtio_cfg.device_activated {
+            return Err(Error::AlreadyActivated);
+        }
+
+        if !self.queues_valid() {
+            return Err(Error::QueuesNotValid);
+        }
```

Comment:
While successful activation is a bit overkill to test, we can easily test these two error scenarios (and the device_type function) as follows:
```rust
        let block = Block::new(args).unwrap();
        assert_eq!(block.lock().unwrap().device_type(), BLOCK_DEVICE_ID);

        assert!(matches!(
            block.lock().unwrap().activate(),
            Err(Error::QueuesNotValid)
        ));

        block.lock().unwrap().virtio_config_mut().device_activated = true;
        assert_eq!(block.lock().unwrap().virtio_config().device_activated, true);
        assert!(matches!(
            block.lock().unwrap().activate(),
            Err(Error::AlreadyActivated)
        ));
```